NE-1155: Add v1 API version

alebedev87 opened the pull request:
What's inside:
- New version added in `api/v1/`:
```
operator-sdk create api --group networking.olm --version v1 --kind AWSLoadBalancerController --controller=false --resource=true --namespaced=false
```
- `v1` is the storage version now
- Conversion webhook added
```
operator-sdk create webhook --group networking.olm  --version v1 --kind AWSLoadBalancerController --conversion
```
- `AllNamespaces` install mode added to CSV (required by the conversion webhook)
- Manifests updated (`make manifests`): 
   - CRD got a new new version
   - `v1` samples added
- Code of the controllers is migrated to `v1`
- e2e tests migrated to `v1`
- new e2e test case added to test the conversion from `v1alpha1`

Note:
- The controller is setup to work with `v1` only. This works out even for the objects created before the upgrade to this version of the operator. Because the version which is being watched/used is the [preferred](https://book.kubebuilder.io/multiversion-tutorial/deployment.html#kubectl-and-preferred-versions)(has to do with [the version priority](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority) too), so OpenShift API will take care of making the conversion. This is different from where the objects are stored though. The objects created before the new version was added to CRD are still stored as the previous version until they are touched (patch/update).

alebedev87 commented:
/retest-required

ingvagabund commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Internal []string `json:"internal,omitempty"`
```
Will every internal subnet owned/reconciled by the lb-controller be always tagged with `kubernetes.io/role/internal-elb`? Is that a hardcoded tag? Or, the tag is independent of the lb-controller?

alebedev87 replied:
> Will every internal subnet owned/reconciled by the lb-controller be always tagged with kubernetes.io/role/internal-elb? 

Yes. That's a tab to say that this subnet is a private one, for not internet-facing load balancers.

>  Is that a hardcoded tag? 

Yes, the openshift installer adds these tags. It's also used by the in-tree aws LB controller.

alebedev87 replied:
As discussed, I added more comments to the `status.subnets` to clarify which tags are used to classify the subnets.

ingvagabund commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Public []string `json:"public,omitempty"`
```
Will every public subnet owned/reconciled by the lb-controller be always tagged with `kubernetes.io/role/elb`? Is that a hardcoded tag? Or, the tag is independent of the lb-controller?

alebedev87 replied:
> Will every public subnet owned/reconciled by the lb-controller be always tagged with kubernetes.io/role/elb?

Yes. This is a tag to say that internet facing load balancer can be created here.

> Is that a hardcoded tag?

Yes but this one is not put by the openshift-installer like the private one. For some reason the tagging of the public subnets with ELB role was removed from the installer at some point. We tried to clarify it but didn't manage to change this with the installer team. That's why the operator now does this job.

alebedev87 replied:
As discussed, I added more comments to the `status.subnets` to clarify which tags are used to classify the subnets.

ingvagabund commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	Config *AWSLoadBalancerDeploymentConfig `json:"config,omitempty"`
+
+	// AWSAddon describes the AWS services that can be integrated with
+	// the AWS Load Balancer.
```
From https://github.com/openshift/enhancements/blob/master/enhancements/ingress/aws-load-balancer-operator.md#implementation-detailsnotesconstraints:
```
By default, all the addons will be enabled.
```
Worth mentioning that in the comment. Also, if all addons are enabled by default, should you rather have `DisabledAddons` field instead? Or, instead of enabling all addons by default, should you rather explicitly enumerate all addons in the `EnabledAddons` field?

alebedev87 replied:
The EP is out dated for this setting. We used to have `DisabledAddons` field at first and then changed it to `EnabledAddons`. So, it's the opposite now - all the addons are disabled by default.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// +kubebuilder:default:=Auto
+	// +kubebuilder:validation:Optional
+	// +optional
+	SubnetTagging SubnetTaggingPolicy `json:"subnetTagging,omitempty"`
```
> How does the operator knows which subnet to tag?

It uses `kubernetes.io/cluster/<infra-id>` tag to get the VPC ID of the cluster it's deployed on. Then it lists all the subnets with `kubernetes.io/cluster/<infra-id>` tag, so all the subnets of the cluster. The `infra-id` is retrieved from `Infrastructure` cluster level resource. 

> Is it always deterministic?

Well, with the assumption that the VPC and the subnets are tagged with the cluster ID (`infra-id`): yes. This is the job of the installer, I suppose other components rely on that tag.

> Any chance of accidentally tagging a wrong subnet?

If a subnet will be tagged as belonging to this cluster: yes. Note: the tagging made is something we believe is missing in the installer as it does add [the right ELB tag for the private subnets](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/deploy/subnet_discovery/#private-subnets) of OpenShift cluster but not for the public ones ([right ELB tag for the public subnets](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/deploy/subnet_discovery/#public-subnets)).

>  When a user explicitly tags specific subnets but the policy is set to Auto will the operator notice there are subnets already tagged and skip the tagging?

Yes.

> Or, will the operator tag subnets which were (maybe for a reason, maybe by accident) omitted?

Sorry, I'm not sure I understood that one. What do you mean by "omitted"?

ingvagabund replied:
> When a user explicitly tags specific subnets but the policy is set to Auto will the operator notice there are subnets already tagged and skip the tagging?

I see the questions is ambiguous. I meant:
When a user explicitly tags specific subnets but the policy is set to Auto will the operator notice there are subnets already tagged and skip tagging of the rest (untagged subnets)?

alebedev87 replied:
> By ommitted I mean ignored when the mode was set to manual. Once the mode gets set to Auto, will the operator start acknowledging subnets there were previously untagged?

Yes, they become OK for the operator - no further action will be taken against those subnets.

> When a user explicitly tags specific subnets but the policy is set to Auto will the operator notice there are subnets already tagged and skip tagging of the rest (untagged subnets)?

The operator just wants to ensure all the correct tags are set. So, if the user manually added the correct tag, the operator is fine with it. However all unttaged subnets are a subject to the tagging, so in `Auto` mode they'll be tagged (see [the other comment](https://github.com/openshift/aws-load-balancer-operator/pull/93#discussion_r1154115781) which details on the transition).

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// controller (default []).
+	//
+	// This value is required so that this controller can function as expected
+	// in parallel to openshift-router.
```
As i mentioned in [the other comment](https://github.com/openshift/aws-load-balancer-operator/pull/93#discussion_r1147326914), this phrase was supposed to be in `IngressClass` field. I will move it there and modify along the way to be clearer.

deads2k replied:
If the value is required, why is empty list allowed?  I don't see validation on this field.  Also, we avoid using maps in the API https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps and have regretted past choices (notice you cannot set permissions on configmap data)

Have a look at how this is structured in the openshift config API: https://github.com/openshift/api/pull/864/files

alebedev87 replied:
@deads2k : I changed the type to "array of objects" for `AdditionalResourceTags` field ([commit](https://github.com/openshift/aws-load-balancer-operator/pull/93/commits/8cf9fcc7b824b97976dab3755fad7d06fc36135c#diff-9e5e37593c04cfcbf496370f1de25e8dc24933c841c471de7a503605949749e9)). Let me know whether it doesn't matche your suggestion.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// +kubebuilder:default:=alb
+	// +kubebuilder:validation:Optional
+	// +optional
+	IngressClass string `json:"ingressClass,omitempty"`
```
> Who owns the created Ingress class?

It's owner referenced to the `AWSLoadBalancerController` resource.

> When the Ingress class is already created, does the ownership get transferred under the lb-controller?

No. Do you think it needs to be adopted?

> Who is responsible for cleaning the Ingress class once it is no longer needed?

It's "owner referenced" to the `AWSLoadBalancerController` resource, so will be deleted with it.

> Is the Ingress class tagged with specific tags as well?

No. Anything which is related to the tags in this API are related to AWS resources.

ingvagabund replied:
> No. Do you think it needs to be adopted?

Not necessarily. The controller should delete only resources it created and ignore already existing.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// +kubebuilder:default:=2
+	// +kubebuilder:validation:Optional
+	// +optional
+	Replicas int32 `json:"replicas,omitempty"`
```
> Is 0 a valid number?

Hm. The managed aws-lb-controller exposes a validating webhook for some customer resources it has, e.g. `IngressClassParams`. Setting replicas to 0 would break the updates of these CRs. However I can imagine a theoretical operational need in disabling the aws-lb-controller for good. So, I tend to answer: _yes_. Why?

> Maybe uint32 instead of int32 to skip validation for negative integers?

`unit32` maybe more accurate indeed. Would it need a conversion webhook (from `v1alpha1` to `v1beta1`) then? or such conversions are handled auto magically?

ingvagabund replied:
No objection here wrt. the signed type

alebedev87 replied:
> Will the operator starts reporting Degraded when Replicas=0? Or, will the operator just starts logging errors about invalid value of Replicas?

Neither of these two. Actually the operator silently ignores the updates to `0`. I agree with you, it's better to make this part of the API by setting the minimum to `1`.

alebedev87 replied:
`+kubebuilder:validation:Minimum:=1` added.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// SubnetTagging indicates the current status of the subnet tags
+	// +kubebuilder:validation:Optional
+	// +optional
+	SubnetTagging SubnetTaggingPolicy `json:"subnetTagging,omitempty"`
```
Yes. We try to do some best effort logic to handle the passage from Auto to Manual (removing the tags from the subnets) and vice versa (tagging the subnets).

ingvagabund replied:
So when a user accidentally switches from Manual to Auto and back to Manual, will the original tagging of subnets before switching to Auto be respected? Or, will all manually tagged subnets be cleared as well?

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Config *AWSLoadBalancerDeploymentConfig `json:"config,omitempty"`
```
> Are there any other deployment spec specific customizations considered for future releases?

Well, we started small to not over complicate the API. I can imagine some classical things like tains/tolerations, resources.

> Have you also considered separating the operator specific fields (e.g. AWSLoadBalancerDeploymentConfig) and ALB specific fields (probably everything else) under separate objects?

Not really. We considered `Config` field as a setting of the managed controller, not operator: how many replicas of the controller we need.

> Right now it seems AWSLoadBalancerControllerSpec represents both spec/status of the operator and spec/status of the ALB.

Well, I think all the fields are consistent enough in terms of their affiliation to the controller. The config thing: I explained our point of view above; the subnet tagging: we have only 1 `AWSLoadBalancerController` for the cluster, so tagging the VPC of this cluster is kinda fine; the addons and additional tags are flags for the aws-lb-controller.

ingvagabund replied:
Thanks for the explanation. I understand it better now. Wrt. config field, as long as the config is concerned about the controller workload (e.g. number of replicas, scheduling constraints, annotations, etc.) it is better to keep it under the same type. I accidentally mixed the operator and the controller bits together when interpreting the type.

alebedev87 replied:
> I accidentally mixed the operator and the controller bits together when interpreting the type.

[This comment](https://github.com/openshift/aws-load-balancer-operator/pull/93#discussion_r1156092316) explains why the subnet tagging (the field that I guess confused you) was put under the controller API.

deads2k commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+// SecretReference contains the information to let you locate the desired secret.
+// Secret is required to be in the operator namespace.
```
Is the operator namespace a const or is it user configurable?  if user configurable, why?

alebedev87 replied:
It's user configurable by via OLM, not via API. However it's a big doubt that the operator will work properly in any other namespace than [the suggested one](https://github.com/openshift/aws-load-balancer-operator/blob/main/bundle/manifests/aws-load-balancer-operator.clusterserviceversion.yaml#L60).

deads2k replied:
> The OLM has the installation namespace field in the UI, so nothing prevents the user from installing into a wrong namespace. We did our best to talk the user out of doing so.

You could crashloop your pod using the downward API to inspect the namespace, right?

deads2k commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	Credentials *SecretReference `json:"credentials,omitempty"`
+}
+
+type AWSLoadBalancerDeploymentConfig struct {
```
How do you decide which fields are customizeable?  For instance, would a tolerations be acceptable?  env var specification?  labels?

deads2k commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	EnabledAddons []AWSAddon `json:"enabledAddons,omitempty"` // indicates which AWS addons should be disabled.
+
+	// Credentials is a reference to a secret containing
+	// the AWS credentials to be used by the controller.
```
What data needs to be contained in the secret being referenced?  Which data key is it under.  What format is the value stored in?

alebedev87 replied:
AWS credentials in the format of an AWS credentials file: static (aws key + secret) or temporary/sts (role arn + token file).

alebedev87 replied:
Updated, let me know if it still needs some clarifications.

deads2k commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// ObservedGeneration is the most recent generation observed.
```
observed or applied?  When this matches metadata.generation, will the change on the deployment be complete or in progress?

alebedev87 replied:
> When this matches metadata.generation, will the change on the deployment be complete or in progress?

The change the operator made? Yes but the deployment may still be changing its state though . We check for the available and updated replicas from the deployment status to update our CR's status.

Also the same `Condition` type is used for the secret which is a little less under this operator's control. As the secret may be provisioned by the cloud credentials operator. I think for the secret case `Observed` makes even more sense as here the operator waits for the other operator's job to complete.

deads2k commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// will be provisioned and have the required resource tags on them. Whereas when
+	// set to manual, this responsibility lies on the user.
+	//
+	// +kubebuilder:default:=Auto
```
What transitions are allowed?  Auto to Manual?  Manual to Auto?

alebedev87 replied:
I [described in details](https://github.com/openshift/aws-load-balancer-operator/pull/93#discussion_r1154115781) what happens in the transitions. Also I updated the comment to explain the transitions to the user.

deads2k commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// +optional
+	AdditionalResourceTags map[string]string `json:"additionalResourceTags,omitempty"`
+
+	// IngressClass specifies the Ingress class which the controller will reconcile.
```
does this take effect when the setting is Manual?

alebedev87 replied:
`AdditionalResourceTags` field or `IngressClass`? (GitHub rendered the comment a little misleading)

`AdditionalResourceTags` are the tags to be added to the resources created by the operand, they are added by the operand itself.
`Auto` or `Manual` modes are here to control the tagging of the subnets which are [per-requisites](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/deploy/subnet_discovery/#subnet-auto-discovery) for the operand to work correctly. Those tags are added by the operator, for the public subnets mostly as the private ones are properly tagged by the openshift installer.

alebedev87 replied:
As discussed, there was a confusion between the subnet tagging done by the operator and the tagging of the load balancers done by the managed controller.

deads2k commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	EnabledAddons []AWSAddon `json:"enabledAddons,omitempty"` // indicates which AWS addons should be disabled.
```
What happens when an item is removed from teh list?

alebedev87 replied:
Each item in this list is 1-1 mapping to a boolean addon flag in the operand. All the addons flags are set to `false` by default and reset to `true` only for the items mentioned in the list. So, if an item is removed the corresponding flag will be set back to the default: `false`.

alebedev87 replied:
Update the comment to clarify the behavior (and the why) of the enabling and disabling of the addons.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	EnabledAddons []AWSAddon `json:"enabledAddons,omitempty"` // indicates which AWS addons should be disabled.
```
The current: no; they all get mapped to boolean flags of the operand:  `--enable-waf`, `--enabled-wafv2`, `--enable-shield`.  There are no other addons for the moment but I suppose that the upstream will try to remain consistent about the future addons to follow the same boolean kind of style.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// +optional
+	SubnetTagging SubnetTaggingPolicy `json:"subnetTagging,omitempty"`
+
+	// Default AWS Tags that will be applied to all AWS resources managed by this
```
Done for this one and for the other fields too.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	EnabledAddons []AWSAddon `json:"enabledAddons,omitempty"` // indicates which AWS addons should be disabled.
```
Yes, the EP is right here. It describes the current behavior of the upstream controller this operator manages. There is not much we can do here, we just provided the API knobs to mimic the behavior.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Subnets *AWSLoadBalancerControllerStatusSubnets `json:"subnets,omitempty"`
```
> Does nil Subnets has any special meaning besides saving some memory?

Yes, it's an easy way for the operator [to check whether the subnets were tagged or not](https://github.com/alebedev87/aws-load-balancer-operator/blob/main/pkg/controllers/awsloadbalancercontroller/controller.go#L105-L106).

deads2k replied:
>Yes, it's an easy way for the operator to check whether the subnets were tagged or not.

why would a length check not do the same?

alebedev87 replied:
> why would a length check not do the same?

It's a compound `struct`, we could have came up with some logic to decide how to calculate the "untagged" state from the struct's fields. However the pointer does this job easy and clear.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// +kubebuilder:default:={}
+	// +kubebuilder:validation:Optional
+	// +optional
+	AdditionalResourceTags map[string]string `json:"additionalResourceTags,omitempty"`
```
> And if removal happens (which seems natural for a user to expect by default), how does the controller know which tags to remove?

This field is applied on the AWS resources which the controller owns (e.g. load balancers). So it's able to calculate the desired state for them, including the desired tags which partially (others are hard coded) come from this field. The removal of the tags as well as updates will be propagated to the AWS resources.
I updated the comment to mention this.

deads2k replied:
> This field is applied on the AWS resources which the controller owns (e.g. load balancers). So the controller is able to calculate the desired state for them, including the desired tags which partially (others are hard coded) come from this field. The removal of the tags as well as updates will be propagated to the AWS resources.
I updated the comment to mention this.

Does the operator remove any unknown tags?  If a user applies a different tag, does the operator allow that tag or remove it?

alebedev87 replied:
> Does the operator remove any unknown tags?

It's the managed controller who's responsible for the additional tags. But yes, the controller will remove any unsolicited tags as it owns the tagging on the AWS resources it creates.

> If a user applies a different tag, does the operator allow that tag or remove it?

Remove it.

alebedev87 commented:
Updated the descriptions for `additionalResourceTags` and `enableAddons`.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	// +kubebuilder:default:=2
+	// +kubebuilder:validation:Optional
+	// +optional
+	Replicas int32 `json:"replicas,omitempty"`
```
Well, indeed, we didn't think of HPA at the time of writing of the operator. So, it'll be fighting with HPA.
How hard the requirement of being "HPA friendly" is?

deads2k replied:
I was thinking about this in terms of which fields should be customizeable and how that would impact an API here.  I don't object per se, but deciding where you limit is worthwhile.  For instance, if this is the only value you want to allow customization for, there's no reason to nest it since .spec.replicas is common if not canonical.

alebedev87 replied:
I see, yes for the moment we agreed on a single customization value - replicas. I believe we considered moving it up in the API  but based on the experience with the other operators where it was quite hard to argue with the customer about certain customizations (mostly scheduling: tolerations, nodeSelector, etc) we thought that this field would make sense to be a dedicated struct.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
```
Done.

alebedev87 commented:
The patch strategy markers are added to status' `Condition`s.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+
+// SecretReference contains the information to let you locate the desired secret.
+// Secret is required to be in the operator namespace.
+type SecretReference struct {
```
There was a similar (not entirely though) [discussion](https://github.com/openshift/aws-load-balancer-operator/pull/85#discussion_r1038628451) we had at the time I added the field. I guess my arguments didn't grow in number since then :)  I only have a future extensibility as a reasoning for a specific type. However since it's already a second time I'm suggested to use an existing type, I probably can go on an use `LocalSecretReference` or `SecretNameReference`. Especially if we'll do some other changes which would force me to add a conversion webhook.

alebedev87 replied:
I'm wondering if the wording `Local` can be misleading. `AWSLoadBalancerController` CR is cluster scope, so it's not quite clear that `Local` refers to `LocalToOperator`. Maybe something like `SecretNameReference` would do the thing?

deads2k replied:
Accepted, though unpredictable namespaces are a pain.

ingvagabund commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,249 @@
+	//
+	// When in "Auto", the operator will detect the subnets where the load balancers
+	// will be provisioned and have the required resource tags on them. Whereas when
+	// set to "Manual", this responsibility lies on the user. The subnets tagged by the operator
```
Will the subnets or the subnet tags be removed?

alebedev87 replied:
Well seen! Updated the comment to mention the subnet tags, not subnets.

alebedev87 commented:
Addressed [the comment](https://github.com/openshift/aws-load-balancer-operator/pull/93#discussion_r1154460277) and added a conversion webhook. The latter will break the installation of the operator in CI because CSV is now in the `AllNamespace` install mode (required for the conversion webhook).

ingvagabund commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,206 @@
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	IngressClass string `json:"ingressClass,omitempty"`
```
Is it guaranteed you will always default to `alb`? Or, can the default value change over time? Preferably, the default 
value should be set in here so it is part of the API. Instead of relying on the operator implementation to perform the defaulting.

alebedev87 replied:
> EDIT: I just noticed you already define the default value in the spec part.

Yes, the default to `alb` is part of the API.

ingvagabund commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,248 @@
+	// subnetTagging indicates the current status of the subnet tags
+	// +kubebuilder:validation:Optional
+	// +optional
+	SubnetTagging SubnetTaggingPolicy `json:"subnetTagging,omitempty"`
```
Is `SubnetTagging` field in the status necessary given there's nothing to reconcile in the cluster wrt. the tagging policy? When `.spec.subnetTagging` is set, the operator starts operating/tagging over the subnets. What additional value is `.spec.subnetTagging` adding? Is `.status.subnetTagging` changed right away after `.spec.subnetTagging` is? Or, is there some additional work/activity performed right after `.spec.subnetTagging` changes but before `.status.subnetTagging` gets updated?

alebedev87 replied:
> Is .status.subnetTagging changed right away after .spec.subnetTagging is?

No.

> Or, is there some additional work/activity performed right after .spec.subnetTagging changes but before .status.subnetTagging gets updated?

Yes, the subnet tagging may happen and if it succeeds the status gets set to the same value as in `.spec.subnetTagging`.

> Is SubnetTagging field in the status necessary given there's nothing to reconcile in the cluster wrt. the tagging policy?

Well, it's there to let the user know that their request of the policy changing has been successfully applied.
 
> When .spec.subnetTagging is set, the operator starts operating/tagging over the subnets. What additional value is .spec.subnetTagging adding?

I would say that's the whole value - make sure the required subnets are properly tagged if in `Auto` and let the user do the tagging by themselves if in `Manual`.

ingvagabund commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,248 @@
+)
+
+// AWSLoadBalancerControllerSpec defines the desired state of AWSLoadBalancerController
+type AWSLoadBalancerControllerSpec struct {
```
The description mentions `AWSLoadBalancerControllerSpec defines the desired state of AWSLoadBalancerController` even though `SubnetTagging` policy is related only to the operator. Using `AWSLoadBalancerSpec` would be closer to the intent. Though, given the type is already named as `awsloadbalancercontroller`, the renaming might create some confusion.

alebedev87 replied:
I see what you mean. Yeah, the subnet tagging  belongs to all the instances of `AWSLoadBalancerController`. I don't remember we ever discussed the moving it away from the API because:
1) we are still sort of consistent because we can have only one`AWSLoadBalancerController` instance for the whole cluster (upstream limitation + we don't see a use case for it).
2) we liked the fact that we could use the status of the controller to clearly show the user which subnets are tagged/untagged/public/private. Moving this logic out of the API would make this reporting more difficult.

I think that we would have to revisit the subnet tagging if the need of having more than 1 instance of the controller comes up. However for now, a move like this may be a little too heavy to take especially given that we have the 17th of April as the deadline for the GA release of this operator.

alebedev87 commented:
Addressed [this comment](https://github.com/openshift/aws-load-balancer-operator/pull/93#discussion_r1155898649).

alebedev87 commented:
Addressed [this](https://github.com/openshift/aws-load-balancer-operator/pull/93#discussion_r1156414373) and [this](https://github.com/openshift/aws-load-balancer-operator/pull/93#discussion_r1156414441) comments.

alebedev87 commented:
Consume the target namespace from the current namespace, not OLM generated `targetNamespaces`. The later is set to all namespaces due to the `AllNamespaces` install mode.

alebedev87 commented:
/retest

alebedev87 commented:
`stable-v1` and `stable-v1.0` OLM channels added, to have a single PR for all things `v1beta1`.

alebedev87 commented:
/test e2e-aws-operator

Cluster creation failed.

alebedev87 commented:
/test e2e-aws-rosa-operator

Deprovisioning failed on the role deletion.

alebedev87 commented:
/test e2e-aws-rosa-operator

deads2k commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,256 @@
+	// +optional
+	Subnets *AWSLoadBalancerControllerStatusSubnets `json:"subnets,omitempty"`
+
+	// ingressClass is the current default Ingress class.
```
If the default is fixed, why is there a status field about that default?

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,256 @@
+	// enabledAddons describes the AWS services that can be integrated with
+	// the AWS Load Balancers created by the controller.
+	// Note that when an addon which was previously enabled is disabled
+	// the controller does not remove the existing addon attachment for the provisioned load balancers.
```
Updated the comment to better explain the behavior of the addon enabling/disabling.

alebedev87 commented on api/v1beta1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,256 @@
+	// +optional
+	Public []string `json:"public,omitempty"`
+
+	// tagged is the list of subnet ids which have been tagged by the operator.
```
Yes. If the policy is `Auto`, the operator owns the tagging and will try to reapply the required tags if they are missing.

alebedev87 replied:
I updated the comment for `.spec.subnetTagging` to better explain the fact that in `Auto` mode the operator owns the tagging and will continuously ensure them.

alebedev87 commented:
[This comment](https://github.com/openshift/aws-load-balancer-operator/pull/93#discussion_r1164133143) addressed.

alebedev87 commented on PROJECT:
```diff
@@ -19,6 +19,9 @@ resources:
   group: networking.olm
   kind: AWSLoadBalancerController
-  path: github.com/openshift/aws-load-balancer-operator/api/v1
+  path: github.com/openshift/aws-load-balancer-operator/api/v1beta1
```
Wrong rebase, 2 commits were modifying this file. Corrected.

deads2k commented:
Thanks for the updates

/approve
/lgtm
/hold

holding in case you want another team member for a second look.

alebedev87 commented:
/assign @Miciah

alebedev87 commented on api/v1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,250 @@
+// AWSLoadBalancerControllerSpec defines the desired state of AWSLoadBalancerController
+type AWSLoadBalancerControllerSpec struct {
+
+	// subnetTagging describes how the subnet tagging will be done by the operator.
+	// When in "Auto", the operator will detect the subnets where the load balancers
+	// will be provisioned and will continuously ensure the required resource tags on them.
+	// Whereas when set to "Manual", this responsibility lies on the user. The tags added by the operator
+	// will be removed when transitioning from "Auto" to "Manual". Whereas the tags added by the user
+	// will be left intact when transitioning from "Manual" to "Auto". For more info about the subnet detection
+	// and the required subnet tags see https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/deploy/subnet_discovery.
```
Yeah, it's a pity the things like defaults and enum fields are not displayed by `oc explain` because they are part of the CRD. We have to duplicate them increasing the possibility of the out dated  info.

alebedev87 commented on api/v1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,250 @@
+	// controller. The addition of new tags as well as the update or removal of the existing tags
+	// will be propagated to the AWS resources. The controller owns all the tags of the managed AWS resources,
+	// unsolicited tags are removed.The controller doesn't watch for changes on AWS, so the removal of the unsolicited
+	// tags can only be triggered by an event coming from OpenShift. AWS supports a maximum of 50 tags per resource.
+	// AWSLoadBalancerController reserves 3 tags for its use, the rest is split in half between the tag annotation
+	// which can be set on the ingress or service and this field. Each tag key must be unique.
```
Added.

alebedev87 commented on api/v1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,250 @@
+	SubnetTagging SubnetTaggingPolicy `json:"subnetTagging,omitempty"`
+
+	// additionalResourceTags are the AWS tags that will be applied to all AWS resources managed by this
+	// controller. The addition of new tags as well as the update or removal of the existing tags
+	// will be propagated to the AWS resources. The controller owns all the tags of the managed AWS resources,
+	// unsolicited tags are removed.The controller doesn't watch for changes on AWS, so the removal of the unsolicited
```
Those are 2 different taggings. The subnet tagging is done by the operator to match the mandatory pre-requisites (ELB roles on the subnets). This tagging is what the managed controller does to the resources it creates: ALB, etc. So, the latter is done in the `owned` manner - the aws lb controller owns these resources and don't accept any unsolicited changes. That's the upstream choice we never really tried to question.

alebedev87 replied:
Suggestion applied.

Miciah replied:
Ah, that makes sense.  A naïve reader might miss the distinction between "operator" and "controller" or the distinction between tagging resources and "managing" them.  Can you add a statement to clarify that "AWS resources managed by this controller" and "managed AWS resources" do _not_ include subnets that the operator tags?

alebedev87 replied:
Added the statement about the subnets.

alebedev87 commented on api/v1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,250 @@
+
+type AWSLoadBalancerDeploymentConfig struct {
+	// replicas is the desired number of the controller replicas.
+	// The controller exposes webhooks for IngressClassParams and TargetGroupBinding custom resources.
```
`AWSLoadBalancerController` CR is not handled by the webhooks exposed by the controller. For the moment, it's only handled by the conversion webhook of the operator.

alebedev87 replied:
`the` added.

alebedev87 commented on api/v1/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "networking.olm.openshift.io", Version: "v1"}
+
+	// SchemeBuilder is used to add go types to the GroupVersionKind scheme
+	SchemeBuilder = &scheme.Builder{GroupVersion: GroupVersion}
```
This is a scaffolded code, weird it didn't have the dots at the end.

alebedev87 commented on config/manifests/kustomization.yaml:
```diff
@@ -8,34 +8,19 @@ resources:
+    version: v1
+    kind: Deployment
+    name: controller-manager
+    namespace: system
```
Yes, it's a placeholder. The kubebuilder scaffolding uses `system` namespace everywhere and then it gets replaced by [this kustomize directive](https://github.com/openshift/aws-load-balancer-operator/blob/39db25de6542a04922abe5d9e1450acac0d5bf12/config/default/kustomization.yaml#L1-L2).

Miciah replied:
I see.  Too bad the scaffolding doesn't use [OpenShift templates](https://docs.openshift.com/container-platform/4.12/rest_api/template_apis/template-template-openshift-io-v1.html), which use explicit syntax for parameter substitution.  * grin *.  I'll refrain from going into a rant about scaffolding and generated files.

alebedev87 commented on config/webhook/service.yaml:
```diff
@@ -0,0 +1,13 @@
+
```
This is a scaffolded manifest, must be a kubebuilder's small fault. Removed.

alebedev87 commented on config/webhook/service.yaml:
```diff
@@ -0,0 +1,13 @@
+  namespace: system
+spec:
+  ports:
+    - port: 443
```
It's autogenerated by the kubebuilder scaffolding (when I added the conversion webhook). So it's not really autoreformated. I like more the first type of the indentation. Changed.

alebedev87 commented on test/e2e/operator_test.go:
```diff
@@ -237,6 +239,80 @@ func TestAWSLoadBalancerControllerWithDefaultIngressClass(t *testing.T) {
+	t.Log("Creating v1alpha1 aws load balancer controller instance with default ingress class, additional resource tags and credentials secret")
+
+	name := types.NamespacedName{Name: "cluster", Namespace: "aws-load-balancer-operator"}
+	// The additional resource tags and the credentials secret are added to ALBC
+	// because they changed in v1beta1.
```
> Hm, were none of the E2E tests testing using additional resource tags before you added this test?

Unfortunately no.

> I hope we remember to keep that coverage for additional resource tags if we ever remove this v1alpha1 test.

Let's followup on this one in another PR.

alebedev87 replied:
Suggestion applied.

alebedev87 commented on hack/add-serving-cert.sh:
```diff
@@ -0,0 +1,51 @@
+[ -z "${secret}" ] && echo "ERROR: --secret flag is required" && exit 1
+[ -z "${namespace}" ] && echo "ERROR: --namespace flag is required" && exit 1
+[ -z "${crd}" ] && echo "ERROR: --crd flag is required" && exit 1
+[ ! -x "$(command -v oc)" ] && { echo "ERROR: oc not found"; exit 1; }
```
No reason, didn't see this inconsistency. Updated.

alebedev87 commented on docs/tutorial.md:
```diff
@@ -10,15 +10,17 @@ with the name `cluster` will be reconciled by the operator.
-    example.org/cost-center: 5113232
-    example.org/security-scope: staging
+  - key: example.org/cost-center
+    value: 5113232
```
Good catch!

alebedev87 commented on api/v1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,250 @@
+	// the AWS Load Balancers created by the controller.
+	// Enabling an addon does not immediately enable the feature on the ingress
+	// resources. Instead, it allows for configuration of the feature through the
+	// annotations. Similarly, removing an addon does not immediately disable the feature
+	// on the ingress resources. Instead, it ignores the configuration of the feature
+	// through the annotations. For more info on the addon annotations see
```
The `enabledAddons` acts on the ability of the controller to see specific (addon) annotations on the ingresses.
Example:
- when the wafv2 addon is enabled - the controller starts reacting to `alb.ingress.kubernetes.io/wafv2-acl-arn` annotation. So if the ingress adds or removes this annotation - the controller will create or remove the "ALB-to-WAF" association.
- when the wafv2 addon is removed from the list - the controller stops reacting to  `alb.ingress.kubernetes.io/wafv2-acl-arn` annotation. So if the ingress adds or removes this annotation - nothing will happen to the ALB.

Miciah replied:
Would the godoc be more accurate, or less accurate, if you removed the word "immediately" or changed it to "necessarily"?

The behavior of aws-load-balancer-controller that you describe seems confusing, undesirable, and changeable; has upstream made a definitive statement that that behavior will never change?  If upstream has made no such promise, we probably don't want to make any promise in the API.  However, we should be clear about what we are or aren't promising.  So I think the promise is this:  
* If you add an addon to the list, the controller may or may not enable it on an existing ALB for an ingress with the annotation to enable that addon.
* If you remove an addon from the list, the controller may or may not remove it from an existing ALB for an ingress with the annotation to enable that addon.

Is that correct?

alebedev87 replied:
Rephrased the description removing "immediately" and adding more emphasis on the difference between the ingress annotations and this enabler option.

alebedev87 commented on config/manifests/kustomization.yaml:
```diff
@@ -8,34 +8,19 @@ resources:
+    namespace: system
+  patch: |-
+    # Remove the manager container's "cert" volumeMount, since OLM will create and mount a set of certs.
+    # Update the indices in this path if adding or removing containers/volumeMounts in the manager's Deployment.
+    - op: remove
+      path: /spec/template/spec/containers/1/volumeMounts/0
+    # Remove the "cert" volume, since OLM will create and mount a set of certs.
+    # Update the indices in this path if adding or removing volumes in the manager's Deployment.
```
I think that there is no best place for this comment. By my experience the order of the containers or volumes may be changed in many places: `config/manager/manager.yaml`, in the kustomize patches, by in the order the kustomize patches are applied. By my experience, if the indexes specified here are wrong, the kustomize error can be quite easily  tracked to the patch which uses the indexes.

alebedev87 commented on api/v1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,250 @@
+	// +optional
+	Public []string `json:"public,omitempty"`
+
+	// tagged is the list of subnet ids which have been tagged by the operator.
```
>  Is the operator really discriminating between subnets that it tagged and subnets that something else tagged?

Yes, it has [a technical tag](https://github.com/openshift/aws-load-balancer-operator/blob/main/pkg/controllers/awsloadbalancercontroller/subnettagging.go#L21) to distinguish between the operator and user tagged subnets.

>  Is there no field that lists subnets that merely have the cluster tag?

No, there is no such field. But the subnet can only get into any of these 4 fields if it's from the current cluster (== have the cluster tag).

Miciah replied:
Is the `networking.olm.openshift.io/albo/tagged` tag documented to make it a formal API, or is it an internal implementation detail for now?  

Could you add "which belong to the cluster and" to the godoc for `Internal`, `Public`, `Tagged`, and `Untagged`?

If I understand correctly, that means that `Internal`, `Public`, and `Untagged` partition the set of subnets that have the cluster tag; is that true?  

If I were naming these fields, I might name them something like `InternalRole`, `PublicRole`, `TaggedByOperator`, and `NoRole`.

alebedev87 replied:
> Is the networking.olm.openshift.io/albo/tagged tag documented to make it a formal API, or is it an internal implementation detail for now?

It's an internal implementation detail for now.

> Could you add "which belong to the cluster and" to the godoc for Internal, Public, Tagged, and Untagged?

I thought the fact that I mention this below is enough to say they belong to the cluster:
```
    // subnets contains details of the cluster subnets, those having `kubernetes.io/cluster/${cluster-name}` tag.
    //  
    // +kubebuilder:validation:Optional
    // +optional
    Subnets *AWSLoadBalancerControllerStatusSubnets `json:"subnets,omitempty"`
```
Btw I did't add this description to `AWSLoadBalancerControllerStatusSubnets` type, will do.

> If I understand correctly, that means that Internal, Public, and Untagged partition the set of subnets that have the cluster tag; is that true?

Yes, they all have the cluster tag.

alebedev87 replied:
Changed the description of `Subnets *AWSLoadBalancerControllerStatusSubnets` and `AWSLoadBalancerControllerStatusSubnets` type to state clearly 1) the usage of this data for the controller needs, 2) belonging to the cluster.

alebedev87 replied:
I just recalled [this comment](https://github.com/openshift/aws-load-balancer-operator/blob/39db25de6542a04922abe5d9e1450acac0d5bf12/pkg/controllers/awsloadbalancercontroller/subnettagging.go#L70-L71) which has a TODO to add a check for the internet gateway to better detect the public subnets.

CFields651 commented:
/label px-approved

lihongan commented:
/label qe-approved

alebedev87 commented:
/hold cancel

LGTM label got wiped, so we don't risk to merge the PR anymore.

Miciah commented on api/v1/awsloadbalancercontroller_types.go:
```diff
@@ -0,0 +1,252 @@
+	// unsolicited tags are removed. The controller doesn't watch for changes on AWS, so the removal of the unsolicited
+	// tags can only be triggered by an event coming from OpenShift. AWS supports a maximum of 50 tags per resource.
+	// AWSLoadBalancerController reserves 3 tags for its use, the rest is split between the tag annotation
+	// which can be set on the ingress and this field: : 23 and 24 accordingly. Each tag key must be unique.
```
```suggestion
	// which can be set on the ingress and this field: 23 and 24, respectively. Each tag key must be unique.
```

Miciah commented:
/lgtm
